src/vm/inlinetracking.cpp

```diff
@@ -418,12 +424,181 @@ void PersistentInlineTrackingMap::Save(DataImage *image, InlineTrackingMap* runt
+{
+    STANDARD_VM_CONTRACT;
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_module));
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_inlineeIndex));
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_inlinersBuffer));
+}
+
+#endif //FEATURE_NATIVE_IMAGE_GENERATION
+
+struct InliningHeader
+{
+    short MajorVersion;
+    short MinorVersion;
+    int SizeOfInlineeIndex;
+};
+
+#ifdef FEATURE_NATIVE_IMAGE_GENERATION
+
+void PersistentInlineTrackingMapR2R::ProcessInlineTrackingEntry(SBuffer *inlinersBuffer, SArray<InlineeRecordR2R> *inlineeIndex, InlineTrackingEntry *entry)
+{
+    STANDARD_VM_CONTRACT;
+    // This call removes duplicates from inliners and makes sure they are sorted
+    entry->SortAndDeduplicate();
+    MethodInModule inlinee = entry->m_inlinee;
+    InlineSArray<MethodInModule, 3> &inliners = entry->m_inliners;
+    COUNT_T tatalInlinersCount = inliners.GetCount();
+    _ASSERTE(tatalInlinersCount > 0);
+
+    // Saving number of inliners
+    NibbleWriter inlinersStream;
+    inlinersStream.WriteEncodedU32(tatalInlinersCount);
+
+    // Saving inliners RIDs, each new RID is represented as an adjustment (diff) to the previous one
+    RID prevMethodRid = 0;
+    for (COUNT_T i = 0; i < tatalInlinersCount; i++)
+    {
+        RID methodRid = RidFromToken(inliners[i].m_methodDef);
+        _ASSERTE(methodRid >= prevMethodRid);
+        inlinersStream.WriteEncodedU32(methodRid - prevMethodRid);
+        prevMethodRid = methodRid;
+    }
+    inlinersStream.Flush();
+
+    // Copy output of NibbleWriter into a big buffer (inlinersBuffer) for inliners
+    // and create an InlineeRecord with correct offset
+    InlineeRecordR2R record(RidFromToken(inlinee.m_methodDef));
+    DWORD inlinersStreamSize;
+    const BYTE *inlinersStreamPtr = (const BYTE *)inlinersStream.GetBlob(&inlinersStreamSize);
+    record.m_offset = inlinersBuffer->GetSize();
+    inlinersBuffer->Insert(inlinersBuffer->End(), SBuffer(SBuffer::Immutable, inlinersStreamPtr, inlinersStreamSize));
+
+    inlineeIndex->Append(record);
+}
+
+void PersistentInlineTrackingMapR2R::Save(SBuffer* pSaveTarget, InlineTrackingMap* runtimeMap)
 {
     STANDARD_VM_CONTRACT;
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_module));
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_inlineeIndex));
-    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMap, m_inlinersBuffer));
+    _ASSERTE(pSaveTarget != NULL);
+    _ASSERTE(runtimeMap != NULL);
+
+    SArray<InlineeRecordR2R> inlineeIndex;
+    SBuffer inlinersBuffer;
+
+    // Sort records from runtimeMap, because we need to make sure 
+    // we save everything in deterministic order. Hashtable iteration is not deterministic.
+    COUNT_T runtimeMapCount = runtimeMap->GetCount();
+    InlineTrackingEntry **inlinees = new InlineTrackingEntry *[runtimeMapCount];
```

Comment:
Looking around at numerous other allocation sites in NGEN/R2R paths I am not seeing any that do NULL checks. It appears the common pattern is "new (image->GetHeap()) InlineTrackingEntry *[runtimeMapCount]" that will throw an out of memory exception automatically on failure to allocate.